### PR TITLE
CustomKernelInstall: skip if kernel is already installed

### LINKED ISSUE #453: CustomKernelInstall: skip if kernel is already installed
Testcase CAPTURE-VHD-BEFORE-TEST fails in installation of custom kernel, if the kernel passed as parameter(-CustomKernel) is already installed  in the VM.
This need to be modified to skip if kernel is already installed and CAPTURE-VHD-BEFORE-TEST should be Passed.

07/17/2019 10:48:53 : [ERROR] Kernel upgrade failed in 1 VMs.
07/17/2019 10:48:53 : [ERROR] Custom Kernel: https://lisbuildstorage04.blob.core.windows.net/kernel/3.10.0-957.21.3.el7/kernel-3.10.0-957.21.3.el7.x86_64.rpm installation FAIL. Aborting tests.

build-CustomKernel Logs:
Web link detected. Downloading https://lisbuildstorage04.blob.core.windows.net/kernel/3.10.0-957.21.3.el7/kernel-3.10.0-957.21.3.el7.x86_64.rpm
Custom Kernel:kernel-3.10.0-957.21.3.el7.x86_64.rpm
Removing packages that do not allow the kernel to be installed
Installing kernel-3.10.0-957.21.3.el7.x86_64.rpm
warning: kernel-3.10.0-957.21.3.el7.x86_64.rpm: Header V3 RSA/SHA256 Signature, key ID fd431d51: NOKEY
Preparing...                          ########################################
	package kernel-3.10.0-957.21.3.el7.x86_64 is already installed

### PR DESCRIPTION
**Pull request by DivyaJyothiKasam**
Fixes #453 

Added a check to skip kernel installation, if customKernel is already installed and returns true to proceed further installation.

### CODE REVIEW

**Comment by DivyaJyothiKasam** [comment]
Case 1 
Installing the same kernel which is already installed in the VM.
CENTOS 7.6
Initial kernel version: 3.10.0-957.21.3.el7.x86_64
CustomKernel: https://eosgpackages.blob.core.windows.net/testpackages/Kernels/kernel-3.10.0-957.21.3.el7.x86_64.rpm
Logs of build-CustomKernel.txt
yum_install wget: Success
Web link detected. Downloading https://eosgpackages.blob.core.windows.net/testpackages/Kernels/kernel-3.10.0-957.21.3.el7.x86_64.rpm
Custom Kernel:kernel-3.10.0-957.21.3.el7.x86_64.rpm
Removing packages that do not allow the kernel to be installed
customKernelFilesUnExpanded kernel-3.10.0-957.21.3.el7.x86_64.rpm
Kernel is already installed. So skipping kernel-3.10.0-957.21.3.el7.x86_64.rpm installation.
CUSTOM_KERNEL_ALREADY_INSTALLED
CUSTOM_KERNEL_SUCCESS

RHEL 7.4
Initial Kernel Version: 3.10.0-693.50.3.el7.x86_64
CustomKernel: https://lisbuildstorage04.blob.core.windows.net/kernel/3.10.0-693.50.3.el7/kernel-3.10.0-693.50.3.el7.x86_64.rpm
Logs of build-CustomKernel.txt
yum_install wget: Success
Web link detected. Downloading https://lisbuildstorage04.blob.core.windows.net/kernel/3.10.0-693.50.3.el7/kernel-3.10.0-693.50.3.el7.x86_64.rpm
Custom Kernel:kernel-3.10.0-693.50.3.el7.x86_64.rpm
Removing packages that do not allow the kernel to be installed
customKernelFilesUnExpanded kernel-3.10.0-693.50.3.el7.x86_64.rpm
Kernel is already installed. So skipping kernel-3.10.0-693.50.3.el7.x86_64.rpm installation.
CUSTOM_KERNEL_ALREADY_INSTALLED
CUSTOM_KERNEL_SUCCESS
RHEL 7.3
Initial Kernel Version: 3.10.0-514.28.1.el7.x86_64
CustomKernel = https://eosgpackages.blob.core.windows.net/testpackages/Kernels/kernel-3.10.0-514.28.1.el7.x86_64.rpm
Logs of build-CustomKernel.txt
yum_install wget: Success
Web link detected. Downloading https://eosgpackages.blob.core.windows.net/testpackages/Kernels/kernel-3.10.0-514.28.1.el7.x86_64.rpm
Custom Kernel:kernel-3.10.0-514.28.1.el7.x86_64.rpm
Removing packages that do not allow the kernel to be installed
customKernelFilesUnExpanded kernel-3.10.0-514.28.1.el7.x86_64.rpm
Kernel is already installed. So skipping kernel-3.10.0-514.28.1.el7.x86_64.rpm installation.
CUSTOM_KERNEL_ALREADY_INSTALLED
CUSTOM_KERNEL_SUCCESS
Case 2:
Installing higher kernel
Initial Kernel Version: 3.10.0-514.28.1.el7.x86_64
CustomKernel = https://eosgpackages.blob.core.windows.net/testpackages/Kernels/kernel-3.10.0-514.63.1.el7.x86_64.rpm
Logs of build-CustomKernel.txt
yum_install wget: Success
Web link detected. Downloading https://eosgpackages.blob.core.windows.net/testpackages/Kernels/kernel-3.10.0-514.63.1.el7.x86_64.rpm
Custom Kernel:kernel-3.10.0-514.63.1.el7.x86_64.rpm
Removing packages that do not allow the kernel to be installed
customKernelFilesUnExpanded kernel-3.10.0-514.63.1.el7.x86_64.rpm
Installing kernel-3.10.0-514.63.1.el7.x86_64.rpm
Preparing...                          ########################################
Updating / installing...
kernel-3.10.0-514.63.1.el7            ########################################
Configuring the correct kernel boot order
Updating grub config: /boot/grub2/grub.cfg
CUSTOM_KERNEL_SUCCESS
Case: Downgrading the kernel
Initial Kernel Version: 3.10.0-514.28.1.el7.x86_64
CustomKernel = https://eosgpackages.blob.core.windows.net/testpackages/Kernels/kernel-3.10.0-514.26.2.el7.x86_64.rpm
Logs of build-CustomKernel.txt
yum_install wget: Success
Web link detected. Downloading https://eosgpackages.blob.core.windows.net/testpackages/Kernels/kernel-3.10.0-514.26.2.el7.x86_64.rpm
Custom Kernel:kernel-3.10.0-514.26.2.el7.x86_64.rpm
Removing packages that do not allow the kernel to be installed
customKernelFilesUnExpanded kernel-3.10.0-514.26.2.el7.x86_64.rpm
Installing kernel-3.10.0-514.26.2.el7.x86_64.rpm
Preparing...                          ########################################
	package kernel-3.10.0-514.28.1.el7.x86_64 (which is newer than kernel-3.10.0-514.26.2.el7.x86_64) is already installed
Configuring the correct kernel boot order
Updating grub config: /boot/grub2/grub.cfg
CUSTOM_KERNEL_FAIL

File: Testscripts/Linux/customKernelInstall.sh [hunk 1]
```diff
@@ -379,6 +379,14 @@ function InstallKernel() {
         esac
         customKernelFilesUnExpanded="${CustomKernel#$LOCAL_FILE_PREFIX}"
         LogMsg "Removing packages that do not allow the kernel to be installed"
+        query_customKernel_install=$(rpm -qp --queryformat='%{Version}-%{Release}.%{Arch}\n' ${customKernelFilesUnExpanded})
```
Comment (juhlee-microsoft) [inline Testscripts/Linux/customKernelInstall.sh]: I don't understand this new logic. There is $isKernelUpgraded variable with new value in CommonFunctions library. Why does it need another evaluation for 'query_customKernel_install'?

**Review by juhlee-microsoft — CHANGES_REQUESTED** [review]

**Comment by LiliDeng** [comment]
@DivyaJyothiKasam You can run without CustomKernel when the initial kernel and upgraded kernel are the same. Could you please let know the scenario for this change? Thanks.

File: Libraries/CommonFunctions.psm1 [hunk 2]
```diff
@@ -618,7 +618,13 @@ function Install-CustomKernel ($CustomKernel, $allVMData, [switch]$RestartAfterU
 									Write-LogInfo "Continuing the tests as default kernel is same as $CustomKernel."
 									$isKernelUpgraded = $true
 								} else {
-									$isKernelUpgraded = $false
+									$status = Run-LinuxCmd -ip $vmData.PublicIP -port $vmData.SSHPort -username $user -password $password  `
```
Comment (iamshital) [inline Libraries/CommonFunctions.psm1]: Do the only cat operation here and check the contents in Powershell.

File: Testscripts/Linux/customKernelInstall.sh (hunk 1 above)
Comment (iamshital) [inline Testscripts/Linux/customKernelInstall.sh]: Change the variable name to something like - kernel_version_of_rpm

**Review by iamshital — CHANGES_REQUESTED** [review]

File: Testscripts/Linux/customKernelInstall.sh (hunk 1 above)
Comment (DivyaJyothiKasam) [reply Testscripts/Linux/customKernelInstall.sh]: The new variable "query_customKernel_install" is added in bash script to get the kernel version from customKernel RPM file and check with the installed kernel version on VM. In commonfunctions, variable $isKernelUpgraded gives the state as true/false for upgrade. 
New logic implemented here is to sucessfully Capture the VHD,if customKernel passed is same as that of kernel version in the VM.

**Review by DivyaJyothiKasam — COMMENTED** [review]
> @DivyaJyothiKasam You can run without CustomKernel when the initial kernel and upgraded kernel are the same. Could you please let know the scenario for this change? Thanks.

Kernel versions available in Azure market place image would be unknown. So we need to handle the case for capturing VHD of VM having same customKernel.

File: Testscripts/Linux/customKernelInstall.sh (hunk 1 above)
Comment (juhlee-microsoft) [reply Testscripts/Linux/customKernelInstall.sh]: Your logic eventually determines if the custom kernel is installed or not. This is not about what version of custom kernel is installed. Please use $isKernelUpgraded

**Review by juhlee-microsoft — CHANGES_REQUESTED** [review]

File: Libraries/CommonFunctions.psm1 [hunk 3]
```diff
@@ -618,7 +618,21 @@ function Install-CustomKernel ($CustomKernel, $allVMData, [switch]$RestartAfterU
...
-									$isKernelUpgraded = $false
+                                    $kernelMatchSkip="CUSTOM_KERNEL_ALREADY_INSTALLED"
+                                    Write-LogInfo "custom kernel skip"
+                                    if ( !(Test-Path -Path "$LogDir\$($job.RoleName)-build-CustomKernel.txt" ) ) {
+                                        Write-LogInfo "enetered"
+                                        Copy-RemoteFiles -download -downloadFrom $job.PublicIP -port $job.SSHPort -files "build-CustomKernel.txt" `
+								                -username $user -password $password -downloadTo $LogDir
+                                        if ( ( Get-Content "$LogDir\build-CustomKernel.txt" ) -imatch $kernelMatchSkip -and $CustomKernel.EndsWith(".rpm") ) {
+                                            $isKernelUpgraded = $true
+                                            Write-LogInfo "divya content path true condition"
```
Comment (johnsongeorge-w) [inline Libraries/CommonFunctions.psm1]: Use meaningful log content. Use of names are not good practice.

File: Libraries/CommonFunctions.psm1 [hunk 4]
```diff
@@ -618,7 +618,21 @@ function Install-CustomKernel ($CustomKernel, $allVMData, [switch]$RestartAfterU
...
+                                        } else {
+                                            $isKernelUpgraded = $false
+                                            Write-LogInfo "divya content path false condition"
```
Comment (johnsongeorge-w) [inline Libraries/CommonFunctions.psm1]: Use meaningful log content. Use of names are not good practice.

File: Libraries/CommonFunctions.psm1 [hunk 5]
```diff
@@ -592,7 +592,7 @@ function Install-CustomKernel ($CustomKernel, $allVMData, [switch]$RestartAfterU
 							if ( ( Get-Content "$LogDir\build-CustomKernel.txt" ) -imatch $kernelMatchSuccess ) {
 								$kernelSuccess += 1
 							}
-							Rename-Item -Path "$LogDir\build-CustomKernel.txt" -NewName "$($job.RoleName)-build-CustomKernel.txt" -Force | Out-Null
+							#Rename-Item -Path "$LogDir\build-CustomKernel.txt" -NewName "$($job.RoleName)-build-CustomKernel.txt" -Force | Out-Null
```
Comment (johnsongeorge-w) [inline Libraries/CommonFunctions.psm1]: Why this line is commented ? Dead codes need not to be maintained.

File: Libraries/CommonFunctions.psm1 [hunk 6]
```diff
@@ -618,7 +618,21 @@ function Install-CustomKernel ($CustomKernel, $allVMData, [switch]$RestartAfterU
 									Write-LogInfo "Continuing the tests as default kernel is same as $CustomKernel."
 									$isKernelUpgraded = $true
 								} else {
-									$isKernelUpgraded = $false
+                                    $kernelMatchSkip="CUSTOM_KERNEL_ALREADY_INSTALLED"
+                                    Write-LogInfo "custom kernel skip"
+                                    if ( !(Test-Path -Path "$LogDir\$($job.RoleName)-build-CustomKernel.txt" ) ) {
+                                        Write-LogInfo "enetered"
```
Comment (johnsongeorge-w) [inline Libraries/CommonFunctions.psm1]: Remove all debug logging.
>>> Write-LogInfo "enetered"

**Review by johnsongeorge-w — CHANGES_REQUESTED** [review]

File: Libraries/CommonFunctions.psm1 [hunk 7]
```diff
@@ -618,7 +617,17 @@ function Install-CustomKernel ($CustomKernel, $allVMData, [switch]$RestartAfterU
...
+                                    if ( !(Test-Path -Path "$LogDir\$($job.RoleName)-build-CustomKernel.txt" ) ) {
+                                        Copy-RemoteFiles -download -downloadFrom $job.PublicIP -port $job.SSHPort -files "build-CustomKernel.txt" `
+								                -username $user -password $password -downloadTo $LogDir
+                                        if ( ( Get-Content "$LogDir\build-CustomKernel.txt" ) -imatch $kernelMatchSkip -and $CustomKernel.EndsWith(".rpm") ) {
+                                            $isKernelUpgraded = $true
+                                        } else {
+                                            $isKernelUpgraded = $false
+                                        }
+                                        Rename-Item -Path "$LogDir\build-CustomKernel.txt" -NewName "$($job.RoleName)-build-CustomKernel.txt" -Force | Out-Null
+                                    }
```
Comment (iamshital) [inline Libraries/CommonFunctions.psm1:633]: Also, handle else condition.

**Review by iamshital — CHANGES_REQUESTED** [review]

File: Testscripts/Linux/customKernelInstall.sh (hunk 1 above)
Comment (DivyaJyothiKasam) [reply Testscripts/Linux/customKernelInstall.sh]: The main motive of PR is to continue VHD creation incase of given customKernel version and kernel version of VM is same.
We need to handle to update the status CUSTOM_KERNEL_ALREADY_INSTALLED/CUSTOM_SUCCESS to log file in shell script else the it returns false as it fails for 		line: 603 if ( $kernelSuccess -eq $jobCount ) condition.

Comment (juhlee-microsoft) [reply Testscripts/Linux/customKernelInstall.sh]: I don't understand. There are 2 changes in a PR. Are they relevant or not? Are you fixing one in the library and the other in the test script? If the change is in the library, can this variable use in the test script?

**Comment by DivyaJyothiKasam** [comment]
closing the PR. Came up with new approach, will raise new PR.

**Pull request closed without merging**